### DISCUSSION TITLE
how to subscribe to ring intercom

### DISCUSSION THREAD

**Discussion by pikooli** [dgreif/ring#1380] [Q&A] [answered]
Hello, I try to make an open source api that would let people define a set of time where the intercom will open the door automatically when someone ring. 

But I don't really understand how to subscribe to ring with the intercom api. 

for now my code look like this : 

```
import "dotenv/config";

import { RingApi, RingIntercom } from "ring-client-api";

const ringApi = new RingApi({
  refreshToken: process.env.RING_REFRESH_TOKEN!,
});
let sub: any;

export async function GET() {
  const locations = await ringApi.getLocations();
  const location = locations[0];
  const intercom = location.intercoms[0];
  if (!sub) {
    sub = await intercom.subscribeToDingEvents().catch((e) => {
      console.log("Failed to subscribe " + e + " to ding events");
    });
  }

  return Response.json("subscribe");
}

export const dynamic = "force-dynamic"; // defaults to auto
```

I look at the api but didn't find a way to pass a action when we receive an alert. 
If I could receive help plz. 

Thanks a lot

ps : I use Nextjs 14.0.4 and "ring-client-api": "^12.1.0"

**Comment by pikooli** [comment, marked as answer]
I rethink it and did something like this : 
```
import { RingApi } from "ring-client-api";

const ringApi = new RingApi({
  refreshToken: process.env.RING_REFRESH_TOKEN!,
});

ringApi.getLocations().then((locations) => {
  const location = locations[0];
  const intercom = location.intercoms[0];
  intercom.onDing.subscribe(() => {
    intercom.unlock();
  });
});

```

I didn't test it yet, will test it when I will have access to a ring intercom.

**Reply by haircrime** [reply]
Hi, did you ever get this to work. 
I triee polling, or live events. Nothing is showing events, when I ring the building door bell.
Cheers
Mario




Polling:

import { RingApi } from 'ring-client-api';
import { promisify } from 'util';
import { randomUUID } from 'crypto';

const setTimeoutPromise = promisify(setTimeout);

/**
 * -------------------------------------------------------------------
 * FINAL VERSION: Polling Strategy
 * -------------------------------------------------------------------
 
 *
 * How to run this script:
 * RING_REFRESH_TOKEN="your_new_refresh_token_goes_here" node xx.mjs
 */
const refreshToken = process.env.RING_REFRESH_TOKEN;
const hardwareId = process.env.RING_HARDWARE_ID || randomUUID();
const POLLING_INTERVAL_MS = 3000; // Check for updates every 3 seconds

// A simple function to compare two objects.
function deepEqual(obj1, obj2) {
  return JSON.stringify(obj1) === JSON.stringify(obj2);
}

async function listenToIntercom() {
  if (!refreshToken) {
    console.error('❌ Error: RING_REFRESH_TOKEN environment variable is not set.');
    process.exit(1);
  }

  console.log('[🚀] Starting Ring Intercom Listener (Polling Mode)...');
  console.log(`[⚙️] Using Hardware ID: ${hardwareId}`);

  try {
    const ringApi = new RingApi({
      refreshToken,
      hardwareId,
      debug: false, // Keep debug off for clean output
    });

    // Use a Map to store the last known state of each intercom, keyed by its ID.
    const deviceStates = new Map();

    console.log('[🔍] Performing initial device discovery...');
    const initialLocations = await ringApi.getLocations();
    let foundAnyIntercoms = false;

    for (const location of initialLocations) {
      console.log(`\n📍 Processing location: "${location.name}"`);
      if (location.intercoms.length === 0) {
        console.log('   -> No intercoms found at this location.');
        continue;
      }

      for (const intercom of location.intercoms) {
        foundAnyIntercoms = true;
        console.log(`   -> 🎧 Found Intercom: ${intercom.name} (ID: ${intercom.id})`);
        console.log(`      Storing initial state...`);
        deviceStates.set(intercom.id, intercom.data); // Store initial state
      }
    }

    if (!foundAnyIntercoms) {
      console.warn('\n⚠️  No Ring Intercom devices were found. Exiting.');
      process.exit(0);
    }
    
    console.log(`\n[✅] Initial state captured. Starting polling loop...`);
    console.log(`    (Polling for changes every ${POLLING_INTERVAL_MS / 1000} seconds. Press Ctrl+C to exit)`);

    // --- GLOBAL POLLING LOGIC ---
    setInterval(async () => {
      try {
        const currentLocations = await ringApi.getLocations();
        for (const location of currentLocations) {
          for (const intercom of location.intercoms) {
            const lastKnownState = deviceStates.get(intercom.id);
            const currentState = intercom.data;

            if (lastKnownState && !deepEqual(currentState, lastKnownState)) {
              console.log('-------------------------------------------------');
              console.log(`🔔 [STATE CHANGE DETECTED] for ${intercom.name}`);
              console.log(`      Time: ${new Date().toLocaleString()}`);
              console.log('      Previous State:', lastKnownState);
              console.log('      New State:', currentState);
              console.log('-------------------------------------------------\n');
              
              deviceStates.set(intercom.id, currentState);
            }
          }
        }
      } catch (e) {
        console.error(`[ERROR] Failed during polling cycle:`, e);
      }
    }, POLLING_INTERVAL_MS);

    // Keep the main script alive so the setInterval can run.
    await setTimeoutPromise(4 * 3600 * 1000); 
    console.log('[⌛] Polling script finished after 4 hours.');

  } catch (error) {
    console.error('\n❌ A fatal error occurred during startup:');
    console.error(error);
    process.exit(1);
  }
}

listenToIntercom();



LIVE Events:
// intercom_listener_v11.mjs
import { RingApi } from 'ring-client-api';
import { promisify } from 'util';
import { randomUUID } from 'crypto';

const setTimeoutPromise = promisify(setTimeout);

/**
 * -------------------------------------------------------------------
 * SENSITIVE CONFIGURATION
 * -------------------------------------------------------------------
 * This version correctly handles multiple locations and multiple intercoms,
 * attaching a listener to every intercom found on the account.
 *
 * How to run this script:
 * RING_REFRESH_TOKEN="your_new_refresh_token_goes_here" node xx.mjs
 */
const refreshToken = process.env.RING_REFRESH_TOKEN;
const hardwareId = process.env.RING_HARDWARE_ID || randomUUID();

async function listenToIntercom() {
  if (!refreshToken) {
    console.error('❌ Error: RING_REFRESH_TOKEN environment variable is not set.');
    process.exit(1);
  }

  console.log('[🚀] Starting Ring Intercom Listener (Multi-Location Mode)...');
  console.log(`[⚙️] Using Hardware ID: ${hardwareId}`);

  try {
    const ringApi = new RingApi({
      refreshToken,
      hardwareId,
      debug: false, 
    });

    console.log('[🔍] Fetching all locations and devices...');
    const locations = await ringApi.getLocations();

    if (locations.length === 0) {
        console.error('❌ No locations found on this account. Exiting.');
        process.exit(1);
    }
    
    let foundAnyIntercoms = false;

    // Loop through every location found on the account.
    for (const location of locations) {
        console.log(`\n📍 Processing location: "${location.name}"`);

        if (location.intercoms.length === 0) {
            console.log('   -> No intercoms found at this location.');
            continue;
        }

        // For the current location, loop through every intercom.
        for (const intercom of location.intercoms) {
            foundAnyIntercoms = true;
            console.log(`   -> 🎧 Found Intercom: ${intercom.name} (ID: ${intercom.id})`);
            console.log('      Attaching `onDing` listener...');

            // Attach the listener to this specific intercom.
            intercom.onDing.subscribe((ding) => {
              console.log('-------------------------------------------------');
              console.log(`🔔🔔🔔 Ding Dong! Event from ${intercom.name} at ${location.name}`);
              console.log(`      Time: ${new Date().toLocaleString()}`);
              console.log('      Event Details:', ding);
              console.log('-------------------------------------------------\n');
            });
        }
    }

    if (!foundAnyIntercoms) {
        console.error('\n❌ No intercom devices were found across any of your locations. Exiting.');
        process.exit(1);
    }

    console.log('\n[✅] Setup complete. Waiting for a "ding" event from any intercom...');
    console.log(`    (Press Ctrl+C to exit)`);
    
    // Keep the main script alive to listen for events.
    await setTimeoutPromise(4 * 3600 * 1000); 
    console.log('[⌛] Listener script finished after 4 hour.');

  } catch (error) {
    console.error('\n❌ A fatal error occurred during startup:');
    console.error(error);
    process.exit(1);
  }
}

listenToIntercom();
